apps/ejabberd/src/mongoose_riak_sup.erl

```diff
@@ -0,0 +1,117 @@
+%%==============================================================================
+%% Copyright 2015 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%==============================================================================
+-module(mongoose_riak_sup).
+
+-behaviour(supervisor).
+
+%% API
+-export([start/2]).
+-export([start_link/2]).
+-export([stop/0]).
+-export([get_riak_pools_count/0]).
+
+%% Supervisor callbacks
+-export([init/1]).
+
+-define(SERVER, ?MODULE).
+
+-include("ejabberd.hrl").
+
+%%%===================================================================
+%%% API functions
+%%%===================================================================
+
+start(Workers, PoolsSpec) ->
+    ChildSpec = {?MODULE, {?MODULE, start_link, [Workers, PoolsSpec]},
+        transient, infinity, supervisor, [?MODULE]},
+    {ok, _} = supervisor:start_child(ejabberd_sup, ChildSpec).
+%%--------------------------------------------------------------------
+%% @doc
+%% Starts the supervisor
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec(start_link(integer(), list()) ->
+    {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link(Workers, PoolSpec) ->
```

Comment:
Should be PoolsSpec instead of PoolSpec. By the way I am not a fan of creating separate pools for every riak node.
I think It should be handled by an external load balancer or DNS server.  I think that the mentioned solution simplifies the logic on our side and allows to easily expand  Riak cluster  - one can add a new riak node and then register it in the load balancer. In the current solution we need to restart/reload MiM nodes and I couple of riak nodes go down the number of active connections will shrink (or even take MiM down?) while the load balancer would just open new connections to the remaining nodes.